envoy/http/filter.h

```diff
@@ -626,6 +641,33 @@ class StreamDecoderFilterCallbacks : public virtual StreamFilterCallbacks {
   virtual absl::optional<absl::string_view> upstreamOverrideHost() const PURE;
 };
 
+/**
```

Comment:
Per below does this need to be part of the public interface, especially the setting part? Can't we just expose the name?

Reply:
> can't this just be internal to the filter manager?

Yes, we can. But it wouldn't make things easier.

As you can see in the current implementation code snippet. The filter manager will create a filter chain for every request by the `FilterChainFactory`. And we can only add filters to the chain ( or filter manager) by the interface of `FilterChainFactoryCallbacks` for now.

```c++
class FilterChainFactory {
public:

  virtual void createFilterChain(FilterChainFactoryCallbacks& callbacks) PURE;

};
```

---

So, yeah we do have some other ways to make the name available for the filter manager. But there will also be some API changes.

First, we can update this `Http::FilterChainFactoryCallbacks` to add some new methods as following snippet. And we can also make the name available by a `FilterChainFactoryCallbacks` wrapper like `FilterChainFactoryCallbacksNameHelper` automatically.

```c++
class FilterChainFactoryCallbacks {
  /**
   * Add a decoder/encoder filter that is used both when reading and writing stream data.
   * @param custom_name supplies the custom configuration name.
   * @param filter_name supplies the canonical filter name.
   * @param filter supplies the filter to add.
   */
  virtual void addStreamFilter(absl::string_view custom_name, absl::string_view filter_name,
                               Http::StreamFilterSharedPtr filter) PURE;

};
```

---

Or we can update the `FilterChainFactory` to expose the factories and names to the filter manager rather than current way. For example, we can create new `FilterManager` API that will handle every named filter factory, then we can handle the name internally in the filter manager.

```c++
// New interface for the filter manager.
class FilterManager {
public:
     void evaluateFilterFactory(absl::string_view custom_name, absl::string_view filter_name,
                               Http::FilterFactoryCb& cb) PURE;
};

class FilterChainFactory {
public:
  // Add filter to the filter manager by the new interface rather than FilterChainFactoryCallbacks directly. Then
  // we can handle the name of filter internally in the filter manager.
  virtual void createFilterChain(FilterManager& filter_manager) PURE;
};
```


Except current implemantion in this PR, I prefer the second way. How do you think about it? cc @mattklein123

Reply:
> So, yeah we do have some other ways to make the name available for the filter manager. But there will also be some API changes.
> 
> First, we can update this `Http::FilterChainFactoryCallbacks` to add some new methods as following snippet. And we can also make the name available by a `FilterChainFactoryCallbacks` wrapper like `FilterChainFactoryCallbacksNameHelper` automatically.
> 
> ```c++
> class FilterChainFactoryCallbacks {
>   /**
>    * Add a decoder/encoder filter that is used both when reading and writing stream data.
>    * @param custom_name supplies the custom configuration name.
>    * @param filter_name supplies the canonical filter name.
>    * @param filter supplies the filter to add.
>    */
>   virtual void addStreamFilter(absl::string_view custom_name, absl::string_view filter_name,
>                                Http::StreamFilterSharedPtr filter) PURE;
> 
> };
> ```

not sure I understand that your second way right. If I understand right, I feel like that is a better choice.

Then we can enable the `ActiveStreamDecoderFilter` or `ActiveStreamFilterBase` to hold the custom name and filter name by this new `addStreamFilter` interface.

Reply:
> not sure I understand that your second way right. If I understand right, I feel like that is a better choice.

> For example, we can create new FilterManager API that will handle every named filter factory, then we can handle the name internally in the filter manager.
```c++
// New interface for the filter manager.
class FilterManager {
public:
     void evaluateFilterFactory(absl::string_view custom_name, absl::string_view filter_name,
                               Http::FilterFactoryCb& cb) PURE;
};

class FilterChainFactory {
public:
  // Add filter to the filter manager by the new interface rather than FilterChainFactoryCallbacks directly. Then
  // we can handle the name of filter internally in the filter manager.
  virtual void createFilterChain(FilterManager& filter_manager) PURE;
};
```

Do you mean this one? I also like this one. If we choose this way finally, I think I would create another seperate PR to update `FilterChainFactory`.

Reply:
haha, sorry, that isn't the one I said. Looks like what I choose is your first way.

It is almost the same as the current version, just change the `addStreamFilter` interface to accept the custom name and filter name, then pass those two names into the `ActiveStreamDecoderFilter` or `ActiveStreamFilterBase`. Then we needn't the filter itself to hold those two names and expose those interfaces.

Reply:
@soulxu In that way, we need to update three different APIs `addStreamFilter`, `addDecoderStreamFilter` and `addEncoderStreamFilter`.

And we need keep two different version (old APIs and new APIs with names) in the `Http::FilterChainFactoryCallbacks` to avoid breaking current filters. (This is the main reason I think this way is not so good.)

Reply:
I think we need to step back a second. I don't really understand why anyone would specify a custom name in code. shouldn't this come from config or the statically known names, and if so, can't that just be done directly by the filter manager? That seems like a much better solution to me. I still don't quite understand why any of this needs to be exposed publicly. It seems it should be all internal to the filter manager and the tracked filters there.

/wait

Reply:
In the current implementation of `FilterChainFactoryCallbacks` and `FilterChainFactory`, there is no any easy way to make the custom/filter names available to the filter manager. There is no position for additional filter info (custom/filter names) in current interfaces.


**We must update stream filter APIs to expose these names**

**or**

**update APIs of `FilterChainFactoryCallbacks` and `FilterChainFactory`.**


In the current PR, I choose to update stream filter APIs to expose these names. And the filter itself can also easily access these name.


The setting part is compromise. By these public setter, we can set the names from configuration or static name uniformly at the outrage of filters. Then we needn't implement the getter one by one. But I admit that public setters really aren't a good design.

Of course, we can step back one second. 
We can choose **update APIs of `FilterChainFactoryCallbacks` and `FilterChainFactory`.**

Reply:
Here is an example filter creation:
```
return [&context, config](Http::FilterChainFactoryCallbacks& callbacks) -> void {
    callbacks.addStreamDecoderFilter(
        std::make_shared<SquashFilter>(config, context.clusterManager()));
  };
```

What if we construct a synthetic FilterChainFactoryCallbacks that we pass to each filter factory. This synthetic callbacks would have the name info. Then, it can trap the add filter calls, add the custom names, and create them internally in the filter manager with the names. Wouldn't that be transparent and pretty clean?

Reply:
Yeah, it is transparent and clean. It's also what I want.

But note that the custom names and filter factories are managed by the HCM config (derived from FilterChainFactory). And when filter manager needs create a filter chain, it will call the creatFilterChain method of FilterChainFactory. And filter manager self will be used as FilterChainFacotryCallbacks parameter of creatFilterChain method.

We can created a wrapper to trap the add filter calls. But the wrapper can only call the APIs exposed by the FilterChainFacotryCallbacks. It cannot make the names available to the filter manager directly.

If we don't want add new APIs in the stream filter, we still need update FilterChainFacotryCallbacks or FilterChainFactory to complete this work.

In fact, I has implemented similar wrapper in this PR.

Reply:
@wbpcode and I discussed this offline and I think came up with a good solution which will involve changing the interface that the filter manager uses but not any of the filters, so there shouldn't be any public changes. @wbpcode do you mind summarizing to make sure everyone likes the plan? Thank you!

Reply:
Get it.